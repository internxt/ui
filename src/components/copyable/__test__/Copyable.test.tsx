import { render, fireEvent, waitFor } from '@testing-library/react';
import { Copyable } from '../';
import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest';
import React from 'react';

describe('Copyable Component', () => {
  beforeEach(() => {
    Object.assign(navigator, {
      clipboard: {
        writeText: vi.fn(),
      },
    });
  });

  const renderCopyable = (props = {}) => {
    return render(<Copyable text="Text to copy" {...props} />);
  };

  afterEach(() => {
    vi.clearAllMocks();
  });

  it('should match snapshot', () => {
    const copyable = renderCopyable();
    expect(copyable).toMatchSnapshot();
  });

  it('should render the component with default props', () => {
    const { getByText, getByRole } = renderCopyable();
    expect(getByText('Text to copy')).toBeInTheDocument();
    expect(getByRole('button')).toBeInTheDocument();
    expect(getByText('Copy to clipboard')).toBeInTheDocument();
  });

  it('should copy text to the clipboard when the button is clicked', async () => {
    const { getByRole } = renderCopyable();
    const button = getByRole('button');
    fireEvent.click(button);

    await waitFor(() => {
      expect(navigator.clipboard.writeText).toHaveBeenCalledWith('Text to copy');
    });
  });

  it('should display "Copied!" tooltip text after clicking the button', async () => {
    const { getByText, getByRole } = renderCopyable();
    const button = getByRole('button');
    fireEvent.click(button);

    await waitFor(() => {
      expect(getByText('Copied!')).toBeInTheDocument();
    });
  });

  it('should disable the button temporarily after copying', async () => {
    const { getByRole } = renderCopyable();
    const button = getByRole('button');
    fireEvent.click(button);

    await waitFor(() => {
      expect(button).toBeDisabled();
    });

    await waitFor(() => {
      expect(button).not.toBeDisabled();
    });
  });

  it('should use custom className and classText when provided', () => {
    const { getByText } = renderCopyable({ className: 'custom-class', classText: 'custom-text-class' });
    const container = getByText('Text to copy').parentElement;
    expect(container).toHaveClass('custom-class');
    expect(getByText('Text to copy')).toHaveClass('custom-text-class');
  });

  it('should use custom tooltip texts when provided', async () => {
    const { getByText, getByRole } = renderCopyable({
      copiedText: 'Custom Copied!',
      copyToClipboardText: 'Custom Copy to clipboard',
    });
    const button = getByRole('button');

    expect(getByText('Custom Copy to clipboard')).toBeInTheDocument();

    fireEvent.click(button);

    await waitFor(() => {
      expect(getByText('Custom Copied!')).toBeInTheDocument();
    });
  });

  it('should render with default className and classText when not provided', () => {
    const { getByText } = renderCopyable();

    const container = getByText('Text to copy').closest('div');
    expect(container).toHaveClass(
      'flex h-11 items-center justify-between rounded-md border border-gray-10 bg-gray-5 px-4',
    );

    const textElement = getByText('Text to copy');
    expect(textElement).toHaveClass('select-text text-gray-80');
  });

  it('should override default className and classText when provided', () => {
    const className = 'custom-container-class';
    const classText = 'custom-text-class';
    const { getByText } = renderCopyable({ className, classText });

    const container = getByText('Text to copy').closest('div');
    expect(container).toHaveClass(className);

    const textElement = getByText('Text to copy');
    expect(textElement).toHaveClass(classText);
  });
});
