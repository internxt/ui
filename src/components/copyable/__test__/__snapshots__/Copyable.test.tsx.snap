// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Copyable Component > should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class=" flex h-11 items-center justify-between rounded-md border border-gray-10 bg-gray-5 px-4"
      >
        <p
          class="select-text text-gray-80"
        >
          Text to copy
        </p>
        <div
          class="relative w-max ml-6"
          style="line-height: 0;"
        >
          <div
            class="pointer-events-none absolute top-full left-1/2 -translate-x-1/2 mt-1.5 flex items-center flex-col-reverse drop-shadow-tooltip transition-all duration-150 scale-50 opacity-0"
          >
            <div
              class="w-max rounded-lg bg-gray-90 px-4 py-1.5 text-center dark:bg-gray-5"
            >
              <p
                class="text-base text-white"
              >
                Copy to clipboard
              </p>
            </div>
            <div
              class="bg-gray-90 dark:bg-gray-5 h-1.5 w-4"
              data-testid="tooltip-arrow"
              style="clip-path: polygon(50% 0%, 0% 100%, 100% 100%);"
            />
          </div>
          <button>
            <svg
              class="shrink-0 text-gray-50 hover:text-gray-60"
              fill="currentColor"
              height="24"
              viewBox="0 0 256 256"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M216,32H88a8,8,0,0,0-8,8V80H40a8,8,0,0,0-8,8V216a8,8,0,0,0,8,8H168a8,8,0,0,0,8-8V176h40a8,8,0,0,0,8-8V40A8,8,0,0,0,216,32ZM160,208H48V96H160Zm48-48H176V88a8,8,0,0,0-8-8H96V48H208Z"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class=" flex h-11 items-center justify-between rounded-md border border-gray-10 bg-gray-5 px-4"
    >
      <p
        class="select-text text-gray-80"
      >
        Text to copy
      </p>
      <div
        class="relative w-max ml-6"
        style="line-height: 0;"
      >
        <div
          class="pointer-events-none absolute top-full left-1/2 -translate-x-1/2 mt-1.5 flex items-center flex-col-reverse drop-shadow-tooltip transition-all duration-150 scale-50 opacity-0"
        >
          <div
            class="w-max rounded-lg bg-gray-90 px-4 py-1.5 text-center dark:bg-gray-5"
          >
            <p
              class="text-base text-white"
            >
              Copy to clipboard
            </p>
          </div>
          <div
            class="bg-gray-90 dark:bg-gray-5 h-1.5 w-4"
            data-testid="tooltip-arrow"
            style="clip-path: polygon(50% 0%, 0% 100%, 100% 100%);"
          />
        </div>
        <button>
          <svg
            class="shrink-0 text-gray-50 hover:text-gray-60"
            fill="currentColor"
            height="24"
            viewBox="0 0 256 256"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M216,32H88a8,8,0,0,0-8,8V80H40a8,8,0,0,0-8,8V216a8,8,0,0,0,8,8H168a8,8,0,0,0,8-8V176h40a8,8,0,0,0,8-8V40A8,8,0,0,0,216,32ZM160,208H48V96H160Zm48-48H176V88a8,8,0,0,0-8-8H96V48H208Z"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
